tests/test_registry.rs

```diff
@@ -0,0 +1,92 @@
+use std::collections::HashMap;
+use std::net::SocketAddr;
+use std::sync::atomic::{AtomicU16, Ordering};
+use std::sync::{Arc, RwLock};
+use std::time::Duration;
+
+use axum::{
+    extract,
+    http::{header, StatusCode},
+    response::IntoResponse,
+    routing::get,
+    Router,
+};
+use bytes::Bytes;
+use miette::{miette, Context as _, IntoDiagnostic};
+
+// in case 4367 is already in use, start with the next port
+static PORT: AtomicU16 = AtomicU16::new(4368);
+
+type State = Arc<RwLock<HashMap<String, Bytes>>>;
+
+/// Run a minimal registry for local testing
+async fn test_registry(listen: SocketAddr) -> miette::Result<()> {
+    let state = Arc::new(RwLock::new(HashMap::<String, Bytes>::new()));
+    let app = Router::new()
+        .route("/*path", get(get_package).put(put_package))
+        .with_state(state);
+    let listener = tokio::net::TcpListener::bind(listen)
+        .await
+        .into_diagnostic()
+        .wrap_err(miette!("failed to listen on {listen:?}"))?;
+    tracing::info!("Listening on {listen:?}");
+    axum::serve(listener, app)
+        .await
+        .into_diagnostic()
+        .wrap_err(miette!("failed to read the token from the user"))
+}
+
+// basic handler that responds with a static string
+async fn get_package(
+    extract::State(state): extract::State<State>,
+    extract::Path(path): extract::Path<String>,
+) -> Result<impl IntoResponse, StatusCode> {
+    tracing::info!("Downloaded package from {path}");
+    let content = state
+        .read()
+        .unwrap()
+        .get(&path)
+        .cloned()
+        .ok_or(StatusCode::NOT_FOUND)?;
+    Ok(([(header::CONTENT_TYPE, "application/x-gzip")], content))
+}
+
+async fn put_package(
+    extract::State(state): extract::State<State>,
+    extract::Path(path): extract::Path<String>,
+    body: Bytes,
+) {
+    tracing::info!("Uploaded package to {path} ({} bytes)", body.len());
+    state.write().unwrap().insert(path, body);
+}
+
+// do not use (flavor = "current_thread") here because the user-provided function is blocking
+#[tokio::main]
+pub async fn with_test_registry<F: FnOnce(&str)>(f: F) {
+    let port = PORT.fetch_add(1, Ordering::Relaxed);
```

Comment:
```
Port 0 can be used by applications when calling the bind() command to request the next available dynamically allocated source port number.
```

You could do SocketAddr::new([127, 0, 0, 1], 0) and get a different free port on each call. You can then call `SocketAddr::port` if you need to know which port was chosen.

Reply:
Oh, right! The code was initially starting a separate process, but now that it's simple `async` code, I can definitely do that.

Done!